File: scripts/aios_mlx_lock.py
"""aios_mlx_lock.py — Cross-process file lock for MLX concurrency coordination.

Mission: prevent the 2026-04-18 OOM cascade pattern where Claude-launched
agents, Task-Queue runners, adhoc scripts, and the vcontext server's own
background loops all hit MLX generate simultaneously. 5x 8GB unified memory
per Qwen3-8B-4bit caller pushes a 36GB box past its swap ceiling and wedges
MLX into uninterruptible sleep.

Contract (compatible with scripts/aios-task-runner.js commit 57115b5
and scripts/aios-mlx-lock.js):

    Lock file: /tmp/aios-mlx-lock
    Format:    "<holder-id>\\n"  (holder-id is ASCII text; trailing newline)
    Stale:     >25min mtime --> auto-cleared (matches task-runner)
    Re-entry: env var AIOS_MLX_LOCK_HOLDER makes nested calls a no-op so
               parent-acquired locks survive child-process acquire attempts.

Usage::

    from aios_mlx_lock import MlxLock, try_mlx_lock, release_mlx_lock

    # Context manager (recommended):
    with MlxLock("locomo-eval:pid-1234", wait_s=1200):
        do_mlx_work()

    # Manual control:
    if try_mlx_lock("locomo-eval:pid-1234", wait_s=5):
        try:
            do_mlx_work()
        finally:
            release_mlx_lock("locomo-eval:pid-1234")

The module is deliberately stdlib-only (no pip install).
"""

from __future__ import annotations

import errno
import os
import time
from pathlib import Path

MLX_LOCK_FILE = "/tmp/aios-mlx-lock"
MLX_LOCK_STALE_S = 25 * 60
MLX_LOCK_DEFAULT_WAIT_S = 20 * 60
MLX_LOCK_ENV_VAR = "AIOS_MLX_LOCK_HOLDER"

_POLL_INTERVAL_S = 0.5


# ── Internal helpers ────────────────────────────────────────────

def _read_holder() -> str | None:
    try:
        return Path(MLX_LOCK_FILE).read_text(encoding="utf-8").strip()
    except (FileNotFoundError, OSError):
        return None


def _is_stale() -> bool:
    try:
        return (time.time() - os.stat(MLX_LOCK_FILE).st_mtime) > MLX_LOCK_STALE_S
    except (FileNotFoundError, OSError):
        return False


def _parent_holds() -> bool:
    """Is the lock held by an ancestor process (via env var)?

    Returns True iff AIOS_MLX_LOCK_HOLDER is set AND the file on disk
    currently contains that exact holder. Subprocess callers then become
    no-ops: they inherit the env var, see the parent still holds it, and
    skip both acquire and release.
    """
    env_holder = os.environ.get(MLX_LOCK_ENV_VAR)
    if not env_holder:
        return False
    return _read_holder() == env_holder


def _try_acquire_atomic(holder_id: str) -> bool:
    """Atomic O_CREAT|O_EXCL|O_WRONLY acquire. True on success."""
    if os.path.exists(MLX_LOCK_FILE):
        if _is_stale():
            try:
                os.unlink(MLX_LOCK_FILE)
            except OSError:
                pass
        else:
            return False
    try:
        fd = os.open(MLX_LOCK_FILE,
                     os.O_CREAT | os.O_EXCL | os.O_WRONLY, 0o644)
    except OSError as e:
        if e.errno == errno.EEXIST:
            return False
        raise
    try:
        os.write(fd, (holder_id + "\n").encode("utf-8"))
    finally:
        os.close(fd)
    return True


# ── Public API ──────────────────────────────────────────────────

def try_mlx_lock(holder_id: str, wait_s: float | None = None) -> bool:
    """Try to acquire the MLX lock, waiting up to wait_s seconds.

    Returns True on success, False on timeout. Re-entrant via
    AIOS_MLX_LOCK_HOLDER env var: if the parent already holds the lock,
    returns True immediately without touching the file.

    Args:
        holder_id: identifying string ("script-name:pid-1234", etc).
                   No newlines; kept under 120 chars.
        wait_s:    max total wait in seconds (default 1200 = 20 min).
    """
    if not isinstance(holder_id, str) or not holder_id or "\n" in holder_id:
        raise ValueError("holder_id must be a non-empty single-line string")
    if _parent_holds():
        return True
    if wait_s is None:
        wait_s = MLX_LOCK_DEFAULT_WAIT_S
    deadline = time.time() + max(0.0, wait_s)
    while True:
        if _try_acquire_atomic(holder_id):
            return True
        if time.time() >= deadline:
            return False
        time.sleep(_POLL_INTERVAL_S)


def release_mlx_lock(holder_id: str) -> None:
    """Release the lock iff we're the holder on disk. Safe no-op otherwise."""
    if _parent_holds():
        return  # parent still holds — parent will release
    try:
        if not os.path.exists(MLX_LOCK_FILE):
            return
        if _read_holder() == holder_id:
            try:
                os.unlink(MLX_LOCK_FILE)
            except OSError:
                pass
    except OSError:
        pass


def mlx_lock_status() -> dict:
    """Read-only peek. Does NOT acquire."""
    try:
        st = os.stat(MLX_LOCK_FILE)
    except FileNotFoundError:
        return {"held": False, "holder": None, "age_s": 0.0, "stale": False}
    age_s = time.time() - st.st_mtime
    stale = age_s > MLX_LOCK_STALE_S
    return {
        "held": not stale,
        "holder": _read_holder(),
        "age_s": age_s,
        "stale": stale,
    }


class MlxLock:
    """Context manager for the MLX file lock.

    Raises TimeoutError if the lock can't be acquired within wait_s.
    No-op if AIOS_MLX_LOCK_HOLDER env marks parent-held.

    Example::

        with MlxLock("locomo-eval:pid-1234", wait_s=1200):
            run_locomo_eval()
    """

    def __init__(self, holder_id: str, wait_s: float | None = None):
        self.holder_id = holder_id
        self.wait_s = wait_s if wait_s is not None else MLX_LOCK_DEFAULT_WAIT_S
        self._acquired = False

    def __enter__(self) -> "MlxLock":
        ok = try_mlx_lock(self.holder_id, wait_s=self.wait_s)
        if not ok:
            holder = _read_holder() or "none"
            raise TimeoutError(
                f"aios-mlx-lock: timeout waiting {self.wait_s}s for MLX "
                f"lock (current holder: {holder})"
            )
        self._acquired = True
        return self

    def __exit__(self, exc_type, exc, tb) -> None:
        if self._acquired:
            release_mlx_lock(self.holder_id)
            self._acquired = False
        # don't suppress exceptions
        return None


# CLI mode: `python3 aios_mlx_lock.py status` — prints JSON status.
if __name__ == "__main__":
    import json
    import sys
    if len(sys.argv) == 2 and sys.argv[1] == "status":
        print(json.dumps(mlx_lock_status()))
        raise SystemExit(0)
    print("Usage: python3 aios_mlx_lock.py status", file=sys.stderr)
    raise SystemExit(2)

File: scripts/aios-mlx-lock.js
// aios-mlx-lock.js — Cross-process file lock for MLX concurrency coordination.
//
// Mission: prevent the 2026-04-18 OOM cascade pattern where Claude-launched
// agents, Task-Queue runners, adhoc scripts, and the vcontext server's own
// background loops all hit MLX generate simultaneously. 5x 8GB unified
// memory per Qwen3-8B-4bit caller pushes a 36GB box past its swap ceiling
// and wedges MLX into uninterruptible sleep.
//
// Contract (compatible with scripts/aios-task-runner.js commit 57115b5):
//   Lock file: /tmp/aios-mlx-lock
//   Format:    "<holder-id>\n"  (holder-id is ASCII text; trailing newline)
//   Stale:     >25min mtime → auto-cleared (matches task-runner)
//   Re-entry: env var AIOS_MLX_LOCK_HOLDER makes nested calls a no-op so
//              parent-acquired locks survive child-process acquire attempts.
//
// Usage:
//   import { withMlxLock, tryMlxLock, releaseMlxLock } from './aios-mlx-lock.js';
//
//   // Fire-and-forget — wraps an async fn, acquires/releases automatically:
//   const result = await withMlxLock('my-script:42', async () => {
//     return await doMlxWork();
//   }, { waitMs: 20 * 60 * 1000 });  // default 20min
//
//   // Manual control (for background loops that need granular release):
//   const ok = await tryMlxLock('my-script:42', { waitMs: 5000 });
//   if (ok) {
//     try { await doMlxWork(); }
//     finally { releaseMlxLock('my-script:42'); }
//   } else {
//     // Another holder timed us out — skip this tick, try again later.
//   }
//
// Safety:
//   - releaseMlxLock only unlinks if the on-disk holder-id matches ours
//     (prevents clobbering someone else's acquired lock on crash-recovery).
//   - AIOS_MLX_LOCK_HOLDER re-entry check is string-exact — nested children
//     inherit the env var and become no-ops; unrelated processes don't.

import { existsSync, statSync, readFileSync, openSync, writeSync, closeSync, unlinkSync } from 'node:fs';

export const MLX_LOCK_FILE = '/tmp/aios-mlx-lock';
export const MLX_LOCK_STALE_MS = 25 * 60 * 1000;
export const MLX_LOCK_DEFAULT_WAIT_MS = 20 * 60 * 1000;
export const MLX_LOCK_ENV_VAR = 'AIOS_MLX_LOCK_HOLDER';

const POLL_INTERVAL_MS = 500;

function _readHolder() {
  try { return readFileSync(MLX_LOCK_FILE, 'utf-8').trim(); }
  catch { return null; }
}

function _isStale() {
  try {
    const st = statSync(MLX_LOCK_FILE);
    return (Date.now() - st.mtimeMs) > MLX_LOCK_STALE_MS;
  } catch { return false; }
}

// Parent already holds the lock? (e.g. task-runner launched this script)
// Check env var AND that the file actually contains that holder (not stale).
function _parentHolds(myHolderId) {
  const envHolder = process.env[MLX_LOCK_ENV_VAR];
  if (!envHolder) return false;
  const onDisk = _readHolder();
  if (!onDisk) return false;
  // Parent held if env matches disk content — our own acquire would recurse.
  // myHolderId is ignored here; even if it differs, re-entry is semantically
  // "this process tree already holds the lock".
  return envHolder === onDisk;
}

// Atomic create — returns true if we successfully acquired, false if EEXIST
// and the existing lock is NOT stale. Clears stale locks on the way.
function _tryAcquireAtomic(holderId) {
  if (existsSync(MLX_LOCK_FILE)) {
    if (_isStale()) {
      try { unlinkSync(MLX_LOCK_FILE); } catch {}
    } else {
      return false;
    }
  }
  // O_CREAT | O_EXCL | O_WRONLY → fails with EEXIST if another caller
  // raced us between the existsSync above and this open. That's fine —
  // the caller's retry loop handles it.
  try {
    const fd = openSync(MLX_LOCK_FILE, 'wx');
    try { writeSync(fd, holderId + '\n'); }
    finally { closeSync(fd); }
    return true;
  } catch (e) {
    if (e && e.code === 'EEXIST') return false;
    // Any other error (permission, disk full) — rethrow so the caller sees it.
    throw e;
  }
}

/**
 * Try to acquire the lock, waiting up to opts.waitMs (default 20min) for it.
 * Returns true on success, false on timeout. Re-entrant: if AIOS_MLX_LOCK_HOLDER
 * matches the on-disk holder, returns true immediately without touching the
 * lock file (release is similarly a no-op in that case).
 *
 * @param {string} holderId  Identifying string ("self-evolve:pid-1234" etc).
 *                           Kept under 120 chars; no newlines.
 * @param {object} opts
 * @param {number} opts.waitMs   Max total wait in ms (default 1,200,000).
 * @param {function} opts.onWait Optional callback(ageMs, holderOnDisk) fired
 *                               once every 60s while waiting, for logging.
 */
export async function tryMlxLock(holderId, opts = {}) {
  if (typeof holderId !== 'string' || !holderId || /\n/.test(holderId)) {
    throw new Error('tryMlxLock: holderId must be a non-empty single-line string');
  }
  if (_parentHolds(holderId)) return true;  // re-entrant no-op
  const waitMs = Math.max(0, opts.waitMs ?? MLX_LOCK_DEFAULT_WAIT_MS);
  const deadline = Date.now() + waitMs;
  let lastLogAt = 0;
  while (true) {
    if (_tryAcquireAtomic(holderId)) return true;
    if (Date.now() >= deadline) return false;
    if (opts.onWait && Date.now() - lastLogAt > 60_000) {
      try { opts.onWait(Date.now() - (Date.now() - 60_000), _readHolder()); } catch {}
      lastLogAt = Date.now();
    }
    await new Promise(r => setTimeout(r, POLL_INTERVAL_MS));
  }
}

/**
 * Release the lock iff we're the holder on disk. No-op if:
 *   - the lock is held by someone else (shouldn't happen in normal flow;
 *     safe behavior if we somehow try to release a lock we don't own)
 *   - AIOS_MLX_LOCK_HOLDER is set (parent holds it; parent will release)
 */
export function releaseMlxLock(holderId) {
  if (_parentHolds(holderId)) return;  // parent still holds — leave alone
  try {
    if (!existsSync(MLX_LOCK_FILE)) return;
    const who = _readHolder();
    if (who === holderId) {
      try { unlinkSync(MLX_LOCK_FILE); } catch {}
    }
  } catch {}
}

/**
 * High-level wrapper: acquire → run fn → release, even on exception.
 * Throws if the lock can't be acquired within opts.waitMs.
 */
export async function withMlxLock(holderId, fn, opts = {}) {
  const acquired = await tryMlxLock(holderId, opts);
  if (!acquired) {
    throw new Error(`aios-mlx-lock: timeout waiting ${opts.waitMs ?? MLX_LOCK_DEFAULT_WAIT_MS}ms for MLX lock (current holder: ${_readHolder() || 'none'})`);
  }
  try { return await fn(); }
  finally { releaseMlxLock(holderId); }
}

/**
 * Lightweight status reader (used by the server's _mlxDrain to decide
 * whether to yield to an external holder). Does NOT acquire.
 * Returns { held: bool, holder: string|null, ageMs: number, stale: bool }.
 */
export function mlxLockStatus() {
  try {
    if (!existsSync(MLX_LOCK_FILE)) return { held: false, holder: null, ageMs: 0, stale: false };
    const st = statSync(MLX_LOCK_FILE);
    const ageMs = Date.now() - st.mtimeMs;
    const stale = ageMs > MLX_LOCK_STALE_MS;
    const holder = _readHolder();
    return { held: !stale, holder, ageMs, stale };
  } catch { return { held: false, holder: null, ageMs: 0, stale: false }; }
}

// CLI mode: `node aios-mlx-lock.js status` — prints JSON status. Useful for
// scripts/tests to check the lock without importing.
if (import.meta.url === `file://${process.argv[1]}`) {
  const arg = process.argv[2];
  if (arg === 'status') {
    process.stdout.write(JSON.stringify(mlxLockStatus()) + '\n');
    process.exit(0);
  }
  process.stderr.write('Usage: node aios-mlx-lock.js status\n');
  process.exit(2);
}
